### PR TITLE
Update CRI-API Capabilities to include a field that allows us to set …

### PR DESCRIPTION
**Pull request by vinayakankugoyal**
…ambient capabilities.

#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
It adds a field to the Capabilities message that allows us to add ambient capabilities to a container. CRI-O and containerd will need to be updated to respect the existence of this field and configure the container accordingly.

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:
go files generated using `hack/update-generated-runtime.sh`

#### Does this PR introduce a user-facing change?

```release-note
Added field add_ambient_capabilities to the Capabilities message in the CRI-API.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/pull/2757
```

### CODE REVIEW

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto [hunk 1]
```diff
@@ -725,6 +725,8 @@ message Capability {
     repeated string add_capabilities = 1;
     // List of capabilities to drop.
     repeated string drop_capabilities = 2;
+    // List of ambient capabilities to add.
```
Comment (tallclair) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:728]: This should document the interactions with `add_capabilities` and `drop_capabilities`. In other words, what happens if a capability is added to multiple lists? Or is that an invalid request?

**Review by tallclair — COMMENTED** [review]
This looks good to me, but you'll need signoff from sig-node

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto (hunk 1 above)
Comment (qiutongs) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:728]: Let's add a link to https://man7.org/linux/man-pages/man7/capabilities.7.html?

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto [hunk 2]
```diff
@@ -725,6 +725,8 @@ message Capability {
     repeated string add_capabilities = 1;
```
Comment (qiutongs) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:725]: Let's explicitly say: `List of capabilities to add to the permitted, inheritable effective and bounding sets.`

Comment (vinayakankugoyal) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:732]: Done.

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto (hunk 1 above)
Comment (vinayakankugoyal) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:735]: Done.

**Review by qiutongs — APPROVED** [review]

**Comment by vinayakankugoyal** [comment]
Gentle ping @yujuhong - Please take a look. Thanks!

**Review by qiutongs — APPROVED** [review]

**Comment by vinayakankugoyal** [comment]
Can someone from /sig-node please review and approve this.

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto [hunk 3]
```diff
@@ -730,11 +730,20 @@ message SELinuxOption {
...
+// inheritable, effective, bounding and ambient sets.
+// If a capability is added to add_capabilities and add_ambient_capabilities lists then it gets added to all sets, i.e.
+// permitted, inheritable, effective, bounding and ambient sets.
 message Capability {
     // List of capabilities to add.
     repeated string add_capabilities = 1;
     // List of capabilities to drop.
     repeated string drop_capabilities = 2;
+    // List of ambient capabilities to add.
+    repeated string add_ambient_capabilities = 3;
```
Comment (SergeyKanzhelev) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:746]: do we have to keep updating the `v1alpha2`? Do we expect this to be implemented on Containerd 1.5, for example?

Comment (vinayakankugoyal) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:746]: I only expect this feature to be implemented in the latest version of containerd. I believe that would be 1.5.x wouldn't it? Sorry I am not very familiar with how CRI-API versions translate to containerd verisons.

Comment (SergeyKanzhelev) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:746]: containerd 1.6 is about to be released. it supports CRI `v1`, which means there is no need to fall back to `v1alpha2`.

Comment (vinayakankugoyal) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:746]: Ack. I'll remove the v1alpha2 changes. Thanks!

Comment (vinayakankugoyal) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto]: Done! PTAL. Thanks for all your help!

Comment (vinayakankugoyal) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto]: @SergeyKanzhelev - looks like removing the changes from v1alpha2 causes `TestV1alpha2ContainerConfig ` test to fail. I think we will need to make the changes in v1 and v1alpha2. WDYT?

**Comment by SergeyKanzhelev** [comment]
> /test pull-kubernetes-unit

This raises an interesting question. On one hand, there is no reason and it confusing to update `v1alpha2`. On other, we are testing the absolute match between v1 and v1alpha2 as we need to ensure there is a way to convert one to another. It comes back to the policy as well: https://github.com/kubernetes/kubernetes/issues/107190 as easiest way just to keep everything at the latest version while the reality is that runtime may not implement the latest version.

Specifically for this PR, I think keeping versions in sync is better. sorry that you need to re-revet it back. I didn't realize this when was commenting [above](https://github.com/kubernetes/kubernetes/pull/104620#discussion_r787060397)

**Comment by vinayakankugoyal** [comment]
> > /test pull-kubernetes-unit
> 
> This raises an interesting question. On one hand, there is no reason and it confusing to update `v1alpha2`. On other, we are testing the absolute match between v1 and v1alpha2 as we need to ensure there is a way to convert one to another. It comes back to the policy as well: #107190 as easiest way just to keep everything at the latest version while the reality is that runtime may not implement the latest version.
> 
> Specifically for this PR, I think keeping versions in sync is better. sorry that you need to re-revet it back. I didn't realize this when was commenting [above](https://github.com/kubernetes/kubernetes/pull/104620#discussion_r787060397)

Thanks added the v1alpha2 changes back!

**Review by mrunalp — APPROVED** [review]

**Pull request merged**